### ISSUE TITLE
Author order not harvesting correctly via crosswalks

### ISSUE THREAD

**Issue by mrdsaunders** [tomwrobel/ora_data_model#181]
Author metadata order in ORA is recorded in `role_order`, which is created by the deposit crosswalk. However, author order is not harvesting from ORA correctly, since the engine cannot sort on <ora:role_order>. The order in the SE ORA record reflects the order of the elements, and not the <ora:role_order> values. From AB:

> I'm afraid the crosswalk engine does not currently have the capability of specifying an adjusted order for multivalued items (like author lists). During deposit, the authors are iterated over in the order they appear, so the <ora:role_order> elements will be in sequence (1, 2, 3, etc). When harvesting back, there is an implicit assumption that the order of the <mods:name> elements in the input document is the same as the ordering of the <ora:role_order>. I take it this is not the case? I hadn't really considered that the incoming data would not be ordered already.

I believe this to be a significant and immediate concern for us, affecting both departmental websites and also REF review, REF acceptance and REF attribution based on contribution. Temporary mitigation by reducing the precedence of the ORA record would have significant implications for the smart report since it would affect OAM data. 

AB has suggested pre-sorting authors prior to the Sword endpoint. Failing that he could investigate amending the engine but this would likely use the upgraded engine that will come with 5.20. 

My initial question is what resource and timescale would be involved in pre-sorting author lists, based on role_order, pre-Sword?

**Comment by mrdsaunders** [comment]
As an aside, given that contributor items deposit in sequence, what is the process by which they get out of order?

**Comment by mrdsaunders** [comment]
Andrew Bennett has offered to correct this temporarily (pending incorporation into the crosswalks engine) in a patch due in about a fortnight.

**Comment by mrdsaunders** [comment]
Just discussing a spec for the fix with Andrew Bennett. He will presort `mods:name` element according to `ora:role_order` element value. 

> An easy approach is to just pre-sort elements with mods:roleTerm equal to Author. But then what about Editors? Do they need to be sorted?

Can I just confirm that ORA reviewers do add a role_order value for editors?

I believe we should ask him to take that into account should there be multiple translators.

**Comment by tomwrobel** [comment]
role order is added to all roles, including editor and translator

**Comment by mrdsaunders** [comment]
So according to the deposit crosswalk, we do send a role_order value for:
Author
Editor
Translator
Contributor
Depositor (not sure how there can be more than 1 though...)

**Comment by mrdsaunders** [comment]
From AB:

> is it possible to have a mods:name element with roleTerms of - say - Author and Editor? In this case, how should the ordering occur? It's very possible I'm over complicating this, and that in pratice none of these complexities will occur, but I'd like to ensure that we've explicitly discussed this!

However unlikely, the SE data model does permit a person to be an author, editor, translator and associated author (contributor) currently. What about in ORA?

**Comment by tomwrobel** [comment]
absolutely normal

A contributor can have multiple roles, each role has a title and a role order. So a person can have:


```
roles:
  - role_title: author
    role_order: 2
  - role_title: editor
    role_order: 3
  - role_title: depositor
    role_order: 1 
```




```

**Comment by mrdsaunders** [comment]
Can there be more than one depositor? I assume so where there are multiple deposits?

**Comment by mrdsaunders** [comment]
But for this matter in question, we would never need to harvest a depositor list back to SE.

**Comment by tomwrobel** [comment]
for us, depositors are ignored beyond the first one

On 16 Mar 2020, at 12:42, David Saunders <notifications@github.com> wrote:

﻿

Can there be more than one depositor? I assume so where there are multiple deposits?

**Comment by mrdsaunders** [comment]
AB has fixed - for 5.19 without patch and then in 5.20. see link to #182
